Optimize GetContinentInstanceId

Gamemechanicwow opened the pull request:
## 🍰 Pullrequest
<!-- Describe the Pullrequest. -->
Optimizes GetContinentInstanceId() by precalculating boundaries needed to do early returns.

Example of existing lines that are used to calculated ContinentInstanceId:
![image](https://github.com/user-attachments/assets/7c6506e1-6d4f-4aa5-8a48-96fcae6bd152)
(Crude mock up. Boundary lines might be off)

Image of orc/troll starting zone, with its own continentinstanceid - show with lines that indicate boundaries and their early return values
![image](https://github.com/user-attachments/assets/dfd88952-4753-4d0f-8086-bbbe6e4b6fd6)

Test show high potential for performance gains
![image](https://github.com/user-attachments/assets/f1c5ee20-0868-4b9b-9309-2ce9cf71e3b6)

**Test explained:**

- loaded all positions from creature table in map 0/1
- made a vector with all positions repeated 50 times
- timed old function
- timed new function
- after timing; a 2nd iteration was done verifying that the new and old implementation returned the same value for all coords in the vector 

**Notes:**
Could use better codeformatting. 
Maybe better naming of structs, vars etc.

Alternative solution
Remove struct and makeboundary - and just add the boundary as constants to function call:
            if (IsNorthTo(x, y, topNorthSouthBoundary, yyy, YYY, xxx, XXX))
                return MAP0_TOP_NORTH;
Upside: less code needed.
Downside: if someone is to edit limits, they also manually need to update the numbers in the function call

### Proof
<!-- Link resources as proof -->
- None

### Issues
<!-- Which Issues does this fix, which are related?
- fixes #XXX
- relates #XXX
-->
- None

### How2Test
<!-- Give a detailed description how to test your PR and confirm it is working as expected.
- Test1
- Test2
-->
- None

### Todo / Checklist
<!-- In case some parts are still missing, important notes, breaking changes and other notable items, list them here. -->
- [ ] Better code formatting
- [ ] Better naming
- [ ] Code comments

Gamemechanicwow committed:
initial

Gamemechanicwow committed:
Update MapManager.cpp

ratkosrb commented:
Why move those to the header file? They are not used outside the file. No need to expose them. Also now if anyone edits the boundaries it will recompile most of the core because it's in a header.

Gamemechanicwow committed:
move constants to .cpp

Gamemechanicwow commented:
Ye, from that perspective I guess it makes sense to keep them in the .cpp file.

I have moved the constants to the .cpp file. Also marked this PR as a draft, since there might be an edgecase with the orgrimmar boundary. It's a small area that doesn't contain any creatures, so my initial test didn't catch it.

Gamemechanicwow committed:
fix north of polyline

Gamemechanicwow commented:
- Added a better check for when a player is further north than the most northen point in the boundary line.
- All poly lines now go from east to west. The first Y coordinate must be less than the last Y coordinate. Added comment and assert to document this new restriction

Diagram showing the 3 types of early outs:

![image](https://github.com/user-attachments/assets/100c9e28-8c41-4e3e-8dec-ef1ffe7ec672)

1 - Player is more south than the most southern point in the boundary line, so the player can't be in this area
2 - Player is to the east or west of the boundary.
3 - Player is above the boundary. We know that boundaries are checked from south to north, so we don't have to worry about the player being in a different boundary. So if the player position is between the starting and end position of the boundary he is in this instance.

ratkosrb commented on src/game/Maps/MapManager.cpp:
```diff
@@ -625,19 +625,218 @@ BattleGroundMap* MapManager::CreateBattleGroundMap(uint32 id, uint32 InstanceId,
+                                        1762.00f, -3746.00f,
+                                        2225.00f, -3596.00f,
+                                        2755.00f, -3766.00f};
+static const InstantiatedContinentBoundary durotarSouthBoundary(MakeBoundary(durotarSouthLimit, sizeof(durotarSouthLimit) / (2 * sizeof(float))));
```
east const

Gamemechanicwow replied:
ye, I should have known 😅 It's fixed now

I made another diagram to remind myself
![image](https://github.com/user-attachments/assets/97e76fed-4531-4c69-8391-0f8d5b58b24b)

Gamemechanicwow committed:
east const
